datastore/warehouse/warehouse.go

```diff
@@ -0,0 +1,278 @@
+package warehouse
+
+import (
+	"compress/gzip"
+	"context"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/blocknative/dreamboat/structs"
+	"github.com/lthibault/log"
+)
+
+var (
+	fileIdleTime   = structs.DurationPerSlot
+	filesPruneTick = fileIdleTime * 3
+	ErrClosed      = errors.New("closed")
+)
+
+type Warehouse struct {
+	logger   log.Logger
+	requests chan StoreRequest
+
+	shutdown       chan struct{}
+	runningWorkers *structs.TimeoutWaitGroup
+
+	m WarehouseMetrics
+}
+
+func NewWarehouse(logger log.Logger, bufSize int) *Warehouse {
+	wh := &Warehouse{logger: logger.WithField("service", "warehouse"), requests: make(chan StoreRequest), runningWorkers: structs.NewTimeoutWaitGroup()} // channel must be unbuffered, for not accepting requests that will not be handled
+	wh.initMetrics()
+	return wh
+}
+
+func (s *Warehouse) RunParallel(ctx context.Context, datadir string, numWorkers int) error {
+	for i := 0; i < numWorkers; i++ {
+		go s.Run(ctx, datadir, i)
+	}
+
+	return nil
+}
+
+func (s *Warehouse) Run(ctx context.Context, datadir string, id int) {
+	s.runningWorkers.Add(1)
+	defer s.runningWorkers.Done()
+
+	logger := s.logger.With(log.F{
+		"workerId": id,
+		"datadir":  datadir,
+	})
+
+	logger.Info("started")
+	defer logger.Info("stopped")
+
+	w := newWorker(id, datadir, logger)
+	ticker := time.NewTicker(filesPruneTick)
+	for {
+		select {
+		case req := <-s.requests:
+			s.handleRequest(ctx, logger, w, req)
+			continue
+		case <-ticker.C:
+			w.closeIdleFiles()
+			continue
+		case <-ctx.Done():
+			w.closeAllFiles()
+			return
+		case <-s.shutdown:
+			w.closeAllFiles()
+			return
+		}
+	}
+}
+
+func (s *Warehouse) handleRequest(ctx context.Context, logger log.Logger, w *worker, req StoreRequest) {
+	file, err := w.getOrCreateFile(req)
+	if err == nil {
+		err = writeToFile(file, req)
+		if err != nil {
+			logger.WithError(err).With(req).Error("failed to write")
+
+			file.Write([]byte("\n"))
+			w.closeFile(file)
+			logger.WithField("filename", file.Name()).Debug("corrupted file closed")
+		}
+	} else {
+		logger.WithError(err).Error("failed to get/create file: %w", err)
+	}
+
+	if err == nil {
+		s.m.Writes.WithLabelValues(toString(req.DataType)).Add(1)
+	} else {
+		s.m.FailedWrites.WithLabelValues(toString(req.DataType)).Add(1)
+	}
+
+	select {
+	case req.err <- err: // if does not block, means it is buffered (1) channel
+	default:
+	}
+
+	select {
+	case req.err <- err:
+		return
+	case <-s.shutdown:
+		logger.WithError(ErrClosed).Error("failed to export request")
+		return
+	case <-ctx.Done():
+		logger.WithError(ctx.Err()).Error("failed to export request")
+		return
+	}
```

Comment:
wouldn't life be easier if you just return, and do whatever is needed one level above? this one is never called asynchronously